starlette/staticfiles.py

```diff
@@ -158,20 +157,25 @@ async def get_response(self, path: str, scope: Scope) -> Response:
         raise HTTPException(status_code=404)
 
     def lookup_path(
-        self, path: str
-    ) -> typing.Tuple[str, typing.Optional[os.stat_result]]:
+        self, path: Path
+    ) -> typing.Tuple[Path, typing.Optional[os.stat_result]]:
         for directory in self.all_directories:
-            full_path = os.path.realpath(os.path.join(directory, path))
-            directory = os.path.realpath(directory)
-            if os.path.commonprefix([full_path, directory]) != directory:
-                # Don't allow misbehaving clients to break out of the static files
-                # directory.
-                continue
+            original_path = Path(directory).joinpath(path)
+            full_path = original_path.resolve()
+            directory = Path(directory).resolve()
             try:
-                return full_path, os.stat(full_path)
+                stat_result = os.lstat(original_path)
+                full_path.relative_to(directory)
+                return full_path, stat_result
+            except ValueError:
+                # Allow clients to break out of the static files directory
+                # if following symlinks.
+                if stat.S_ISLNK(stat_result.st_mode):
+                    stat_result = os.lstat(full_path)
```

Comment:
The `stat_result` from `original_path` and `full_path` is the same, right? We should already have this value from before.

We can also move the `stat_result = os.lstat(original_path)` outside (just before) the `try..except`, right?

Reply:
I think I'm wrong, and this implementation is just fine 👍

Reply:
Actually, I'm not sure... Please check this. As I think what I'm saying is the same as this comment: https://github.com/encode/starlette/pull/1377#discussion_r874622960

Reply:
> The stat_result from original_path and full_path is the same, right?
I don't think so. For the `original_path` in this case will be `S_ISLNK` but when following the link, on the original file it should be `S_ISREG`. I might be wrong though, this is my understanding.

> We can also move the stat_result = os.lstat(original_path) outside (just before) the try..except, right?
I don't think we can, calling `full_path.relative_to(directory)` can cause ValueError, that's why it should be here.

Reply:
I didn't say to put the `relative_to` outside, just the `os.lstat`. But if my statement above is wrong, then this implementation is fine. 👍

Reply:
Oh sorry, calling os.lstat will also raise `FileNotFoundError, NotADirectoryError` exceptions in some cases, that's why it's there. @Kludex